src/System.Private.CoreLib/shared/System/Collections/Concurrent/ConcurrentQueue.cs

```diff
@@ -304,29 +307,63 @@ public int Count
-                            // values and return the sum of the counts of the segments.
                             return GetCount(head, headHead, headTail) + GetCount(tail, tailHead, tailTail);
                         }
                     }
                     else
                     {
-                        // There were more than two segments.  Take the slower path, where we freeze the
-                        // queue and then count the now stable segments.
-                        SnapForObservation(out head, out headHead, out tail, out tailTail);
-                        return unchecked((int)GetCount(head, headHead, tail, tailTail));
+                        // There were more than two segments in the queue.  Fall back to taking the cross-segment lock,
+                        // which will ensure that the head and tail segments we read are stable (since the lock is needed to change them);
+                        // for the two-segment case above, we can simply rely on subsequent comparisons, but for the two+ case, we need
+                        // to be able to trust the internal segments between the head and tail.
+                        lock (_crossSegmentLock)
+                        {
+                            // Now that we hold the lock, re-read the previously captured head and tail segments and head positions.
+                            // If either has changed, start over.
+                            if (head == _head && tail == _tail)
+                            {
+                                // Get the positions from the tail, and as above, if the captured values match the previous reads,
+                                // we can use the values to compute the count of the head and tail segments.
+                                int tailHead = Volatile.Read(ref tail._headAndTail.Head);
+                                int tailTail = Volatile.Read(ref tail._headAndTail.Tail);
+                                if (headHead == Volatile.Read(ref head._headAndTail.Head) &&
+                                    headTail == Volatile.Read(ref head._headAndTail.Tail) &&
+                                    tailHead == Volatile.Read(ref tail._headAndTail.Head) &&
+                                    tailTail == Volatile.Read(ref tail._headAndTail.Tail))
+                                {
+                                    // We got stable values for the head and tail segments, so we can just compute the sizes
+                                    // based on those and add them. Note that this and the below additions to count may overflow: previous
+                                    // implementations allowed that, so we don't check, either, and it is theoretically possible for the
+                                    // queue to store more than int.MaxValue items.
+                                    int count = GetCount(head, headHead, headTail) + GetCount(tail, tailHead, tailTail);
```

Comment:
> GetCount [](start = 84, length = 9)

I am wondering why we don't store the segment's items count inside the segment so we don't have to recalculate it every time?

Reply:
Because it changes on every enqueue and dequeue, which can happen concurrently (multiple of each), and it would then be very expensive to then maintain an accurate count in the segment.

Reply:
that is because we don't take any lock on enqueue/dequeue?

---
In reply to: [189106709](https://github.com/dotnet/coreclr/pull/18035#discussion_r189106709) [](ancestors = 189106709)

Reply:
Yes

Reply:
Should the `head` + `tail` counts be added last after all the internal frozen segments have been counted?

Reply:
*added last  == calculated and added last

Reply:
Why?

Reply:
The head and tail segments aren't frozen so can have stuff added or removed during the evaluation of the count? (e.g. the queue is not completely locked only the intermediate segments or changing segments)

So in the time elapsed during accumulating the intermediate segments counts; the head and tail segments may have changed count? (hence getting them last).

Though I may be misunderstanding

Reply:
i.e it returns a correct point in time count; however it could return a more recent point in time count (e.g. if there were 2000 segments, it would take a slight time to enumerate them all)

Reply:
That makes sense to me.

Reply:
It doesn't matter when the GetCount methods are called; those are deterministic based on the inputs to them, which are captured much earlier on and passed in here.